ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/product/LauncherInfo.java

```diff
@@ -192,4 +234,38 @@ private void writeLinux(String indent, PrintWriter writer) {
 		}
 	}
 
+	@Override
+	public List<IMacBundleUrlType> getMacBundleUrlTypes() {
+		return List.copyOf(fMacBundleUrlTypes.values());
+	}
+
+	@Override
+	public void addMacBundleUrlTypes(List<IMacBundleUrlType> bundleUrlTypes) {
+		IMacBundleUrlType[] addedUrlSchemes = bundleUrlTypes.stream().map(urlType -> {
+			if (urlType != null) {
+				String scheme = urlType.getScheme();
+				if (scheme != null && !fMacBundleUrlTypes.containsKey(scheme)) {
+					urlType.setModel(getModel());
+					fMacBundleUrlTypes.put(scheme, urlType);
+					return urlType;
+				}
+			}
+			return null;
+		}).filter(Objects::nonNull).toArray(IMacBundleUrlType[]::new);
+		if (addedUrlSchemes.length > 0 && isEditable()) {
+			fireStructureChanged(addedUrlSchemes, IModelChangedEvent.INSERT);
+		}
+	}
+
+	@Override
+	public void removeMacBundleUrlTypes(List<IMacBundleUrlType> bundleUrlTypes) {
+		IProductObject[] removedUrlSchemes = bundleUrlTypes.stream() //
+				.map(IMacBundleUrlType::getScheme).map(fMacBundleUrlTypes::remove) //
+				.filter(Objects::nonNull).toArray(IProductObject[]::new);
+
+		if (removedUrlSchemes.length > 0 && isEditable()) {
+			fireStructureChanged(removedUrlSchemes, IModelChangedEvent.REMOVE);
+		}
+	}
```

Comment:
The original logic in this case seems not to be correct since it is attempted twice to remove the scheme from the `fMacBundleUrlTypes` map?
Did you intend to send an event for the list of schemes passed and the list of schemes actually removed?

Reply:
Thanks for catching this! You are right, this was definitely not intended. I accidentally took [`Product#removePlugins()`](https://github.com/eclipse-pde/eclipse.pde/blob/edf6f1ba05f3374428268fb02a8544fbe132859e/ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/product/Product.java#L577-L599) as a template, but that of course has different semantics (removing a plugin also removes its configuration).